node/actors/bft/src/leader/replica_commit.rs

```diff
@@ -118,37 +118,32 @@ impl StateMachine {
 
         // ----------- All checks finished. Now we process the message. --------------
 
-        // TODO: we have a bug here since we don't check whether replicas commit
-        // to the same proposal.
-
         // We add the message to the incrementally-constructed QC.
-        self.commit_qcs
+        let commit_qc = self
+            .commit_qcs
             .entry(message.view.number)
-            .or_insert_with(|| CommitQC::new(message.clone(), self.config.genesis()))
-            .add(&signed_message, self.config.genesis());
+            .or_default()
+            .entry(message.clone())
+            .or_insert_with(|| CommitQC::new(message.clone(), self.config.genesis()));
```

Comment:
we need to protect ourselves from a malicious validator spamming us with many replicacommits for the same view

Reply:
If I understand the code correctly, that should be prevented in the ["Duplicated message" check](https://github.com/matter-labs/era-consensus/blob/bf36b3c2844f742a83a51e553f49c86b11faa78a/node/actors/bft/src/leader/replica_commit.rs#L100), right?

But maybe we can refactor a bit the code to let `commit_qcs` aggregate by view and validator (just like `commit_message_cache`) instead of view and message, and then yes, remove the redundant data structure while performing the same checks. Does it make sense?

Reply:
Ah no, that won't work. I'm aggregating by message purposely to validate the _same_ message gets enough signers' weight

Reply:
ok